File: DevotionLock/Services/Scripture/ScriptureCorpus.swift
//
//  ScriptureCorpus.swift
//  DevotionLock
//
//  Deep module: verified Scripture lookup for Chaplain prefetch + UI cards.
//

import Foundation

struct ChaplainScriptureCitation: Identifiable, Equatable, Hashable, Codable {
    enum Source: String, Codable {
        case bibleAPI = "bible_api"
        case curatedCatalog = "curated_catalog"
    }

    let id: UUID
    let reference: String
    let text: String
    let source: Source
    let bookSlug: String?
    let chapter: Int?
    let startVerse: Int?
    let endVerse: Int?
    let catalogPassageID: String?
    let version: String?

    init(
        id: UUID = UUID(),
        reference: String,
        text: String,
        source: Source,
        bookSlug: String? = nil,
        chapter: Int? = nil,
        startVerse: Int? = nil,
        endVerse: Int? = nil,
        catalogPassageID: String? = nil,
        version: String? = nil
    ) {
        self.id = id
        self.reference = reference
        self.text = text
        self.source = source
        self.bookSlug = bookSlug
        self.chapter = chapter
        self.startVerse = startVerse
        self.endVerse = endVerse
        self.catalogPassageID = catalogPassageID
        self.version = version
    }

    static func fromJSON(_ object: [String: Any]) -> ChaplainScriptureCitation? {
        guard let reference = object["reference"] as? String,
              let text = object["text"] as? String,
              let sourceRaw = object["source"] as? String,
              let source = Source(rawValue: sourceRaw)
        else { return nil }

        return ChaplainScriptureCitation(
            reference: reference,
            text: ScriptureTextNormalizer.dedupeVerseText(text),
            source: source,
            bookSlug: object["book_slug"] as? String,
            chapter: object["chapter"] as? Int,
            startVerse: object["start_verse"] as? Int,
            endVerse: object["end_verse"] as? Int,
            catalogPassageID: object["catalog_passage_id"] as? String,
            version: object["version"] as? String
        )
    }

    var displayText: String {
        ScriptureTextNormalizer.dedupeVerseText(text)
    }
}

enum ScriptureTextNormalizer {
    /// Collapse duplicate pilcrow segments and repeated sentences within verse text.
    static func dedupeVerseText(_ text: String) -> String {
        var normalized = text.trimmingCharacters(in: .whitespacesAndNewlines)
        normalized = normalized.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)

        let pilcrowParts = normalized.components(separatedBy: "¶")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        if pilcrowParts.count > 1 {
            var unique: [String] = []
            for part in pilcrowParts where !unique.contains(part) {
                unique.append(part)
            }
            normalized = unique.joined(separator: " ")
        }

        normalized = normalized.replacingOccurrences(of: #"^¶\s*"#, with: "", options: .regularExpression)

        let sentencePattern = #"[^.!?]+[.!?]+|[^.!?]+$"#
        if let regex = try? NSRegularExpression(pattern: sentencePattern) {
            let range = NSRange(normalized.startIndex..., in: normalized)
            let matches = regex.matches(in: normalized, range: range)
            if matches.count > 1 {
                var uniqueSentences: [String] = []
                for match in matches {
                    guard let swiftRange = Range(match.range, in: normalized) else { continue }
                    let sentence = String(normalized[swiftRange]).trimmingCharacters(in: .whitespacesAndNewlines)
                    guard !sentence.isEmpty, !uniqueSentences.contains(sentence) else { continue }
                    uniqueSentences.append(sentence)
                }
                if !uniqueSentences.isEmpty {
                    normalized = uniqueSentences.joined(separator: " ")
                }
            }
        }

        return normalized.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func joinVerseTexts(_ texts: [String]) -> String {
        dedupeVerseText(texts.joined(separator: " "))
    }
}

enum ScriptureReplyFormatter {
    static func displayText(for message: ChaplainMessage) -> String {
        let base: String
        if message.role == .chaplain, !message.scriptures.isEmpty {
            base = stripCitationText(from: message.text, citations: message.scriptures)
        } else {
            base = message.text
        }
        guard message.role == .chaplain else { return base }
        return ChaplainMessageFormatter.plainText(base)
    }

    private static func stripCitationText(from text: String, citations: [ChaplainScriptureCitation]) -> String {
        var result = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !result.isEmpty else { return result }

        let leadInPattern = #"(?i)^(?:here (?:it|that verse|is that verse|is the verse) is:?|here(?:'s| is) (?:that )?(?:verse|passage|scripture):?)\s*"#
        result = result.replacingOccurrences(of: leadInPattern, with: "", options: .regularExpression)

        for citation in citations {
            let normalized = citation.displayText
            result = stripOccurrence(of: normalized, from: result)

            let unquoted = normalized.trimmingCharacters(in: CharacterSet(charactersIn: "\"'\u{201C}\u{201D}\u{2018}\u{2019}"))
            if unquoted != normalized {
                result = stripOccurrence(of: unquoted, from: result)
            }
        }

        result = result.replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func stripOccurrence(of passage: String, from text: String) -> String {
        guard !passage.isEmpty else { return text }
        var result = text

        let quotedPatterns = [
            #""\#(NSRegularExpression.escapedPattern(for: passage))""#,
            #"'\#(NSRegularExpression.escapedPattern(for: passage))'"#,
            NSRegularExpression.escapedPattern(for: passage),
        ]

        for pattern in quotedPatterns {
            result = result.replacingOccurrences(of: pattern, with: "", options: .regularExpression)
        }

        return result
    }
}

enum ScriptureCorpus {
    private static let discoverLimit = 3

    /// Reference string → exact KJV text (uses on-device Bible API cache when available).
    static func lookup(reference input: String) async -> ChaplainScriptureCitation? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let parsed = BibleReferenceParser.parse(trimmed) else { return nil }

        do {
            let content = try await BibleAPIService.shared.fetchReference(parsed)
            let text = ScriptureTextNormalizer.joinVerseTexts(content.verses.map(\.text))
            guard !text.isEmpty else { return nil }
            let start = content.verses.first?.verseNumber
            let end = content.verses.last?.verseNumber

            return ChaplainScriptureCitation(
                reference: parsed.displayReference,
                text: text,
                source: .bibleAPI,
                bookSlug: content.bookSlug,
                chapter: content.chapter,
                startVerse: start,
                endVerse: end,
                version: content.version == BibleBookCatalog.defaultVersion ? "KJV" : content.version.uppercased()
            )
        } catch {
            return nil
        }
    }

    /// Topic / keyword → curated passages (scripture-first).
    static func discover(query: String, topics: [PassageTopic] = [], limit: Int = discoverLimit) -> [ChaplainScriptureCitation] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        var results = trimmed.isEmpty ? SpiritualPassageCatalog.all : SpiritualPassageCatalog.search(trimmed)

        if !topics.isEmpty {
            results = results.filter { passage in
                topics.contains(where: { passage.topics.contains($0) })
            }
        }

        return results
            .filter { $0.source == .scripture }
            .prefix(limit)
            .map { passage in
                ChaplainScriptureCitation(
                    reference: passage.reference,
                    text: passage.text,
                    source: .curatedCatalog,
                    catalogPassageID: passage.id
                )
            }
    }

    /// Prefetch for obvious references or discovery-shaped questions before sending to Chaplain.
    static func prefetch(for userMessage: String, mood: String) async -> [ChaplainScriptureCitation] {
        let trimmed = userMessage.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        if let parsed = BibleReferenceParser.extractReference(from: trimmed),
           let lookedUp = await lookup(reference: parsed.displayReference) {
            return [lookedUp]
        }

        if let parsed = BibleReferenceParser.parse(trimmed), let lookedUp = await lookup(reference: parsed.displayReference) {
            return [lookedUp]
        }

        if BibleReferenceParser.looksLikeReference(trimmed), let lookedUp = await lookup(reference: trimmed) {
            return [lookedUp]
        }

        let lower = trimmed.lowercased()
        let wantsDiscovery = lower.contains("verse") || lower.contains("passage") || lower.contains("scripture")
            || lower.contains("bible") || lower.contains("find") || lower.contains("about")

        guard wantsDiscovery else { return [] }

        let moodTopics = SpiritualPassageCatalog.matchingTopics(mood: mood, focusTags: TodayFocusStore.tags)
        return discover(query: trimmed, topics: Array(moodTopics))
    }
}

File: supabase/functions/_shared/scripture-corpus.ts
/**
 * ScriptureCorpus — edge implementation for Chaplain tools.
 */

import catalog from "./scripture-catalog.json" with { type: "json" };
import { parseBibleReference } from "./bible-reference-parser.ts";

export interface ScripturePassage {
  reference: string;
  text: string;
  source: "bible_api" | "curated_catalog";
  book_slug?: string;
  chapter?: number;
  start_verse?: number;
  end_verse?: number;
  catalog_passage_id?: string;
  version?: string;
}

const BIBLE_API = "https://cdn.jsdelivr.net/gh/wldeh/bible-api/bibles";
const DEFAULT_VERSION = "en-kjv";

interface BibleVerse {
  book: string;
  chapter: string;
  verse: string;
  text: string;
}

/** Keep first occurrence when the Bible API returns duplicate verse rows. */
function dedupeVersesByNumber(verses: BibleVerse[]): BibleVerse[] {
  const seen = new Set<string>();
  return verses.filter((v) => {
    if (seen.has(v.verse)) return false;
    seen.add(v.verse);
    return true;
  });
}

/** Collapse duplicate pilcrow segments and repeated sentences within joined text. */
export function normalizeScriptureText(text: string): string {
  let normalized = text.trim().replace(/\s+/g, " ");

  const pilcrowParts = normalized.split(/¶/).map((s) => s.trim()).filter(Boolean);
  if (pilcrowParts.length > 1) {
    const unique: string[] = [];
    for (const part of pilcrowParts) {
      if (!unique.some((u) => u === part)) unique.push(part);
    }
    normalized = unique.join(" ");
  }

  normalized = normalized.replace(/^¶\s*/, "");

  const sentences = normalized.match(/[^.!?]+[.!?]+|[^.!?]+$/g)?.map((s) => s.trim()).filter(Boolean) ?? [normalized];
  if (sentences.length > 1) {
    const uniqueSentences: string[] = [];
    for (const sentence of sentences) {
      if (!uniqueSentences.some((u) => u === sentence)) uniqueSentences.push(sentence);
    }
    normalized = uniqueSentences.join(" ");
  }

  return normalized.trim();
}

export async function lookupPassage(reference: string): Promise<ScripturePassage | null> {
  const parsed = parseBibleReference(reference);
  if (!parsed) return null;

  const url = `${BIBLE_API}/${DEFAULT_VERSION}/books/${parsed.bookSlug}/chapters/${parsed.chapter}.json`;
  const res = await fetch(url);
  if (!res.ok) return null;

  const body = await res.json() as { data?: BibleVerse[] };
  const verses = body.data ?? [];
  if (!verses.length) return null;

  let selected = dedupeVersesByNumber(verses);
  if (parsed.verse) {
    const end = parsed.endVerse ?? parsed.verse;
    selected = selected.filter((v) => {
      const n = Number(v.verse);
      return n >= parsed.verse! && n <= end;
    });
    if (!selected.length) selected = dedupeVersesByNumber(verses);
  }

  const text = normalizeScriptureText(selected.map((v) => v.text).join(" "));
  const start = Number(selected[0]?.verse);
  const end = Number(selected[selected.length - 1]?.verse);

  return {
    reference: parsed.displayReference,
    text,
    source: "bible_api",
    book_slug: parsed.bookSlug,
    chapter: parsed.chapter,
    start_verse: Number.isFinite(start) ? start : undefined,
    end_verse: Number.isFinite(end) ? end : undefined,
    version: "KJV",
  };
}

export function discoverPassages(
  query: string,
  topics: string[] = [],
  limit = 3,
): ScripturePassage[] {
  const trimmed = query.trim().toLowerCase();
  let results = catalog.passages as Array<{
    id: string;
    text: string;
    reference: string;
    source: string;
    topics: string[];
  }>;

  if (trimmed) {
    results = results.filter((p) =>
      p.text.toLowerCase().includes(trimmed)
      || p.reference.toLowerCase().includes(trimmed)
      || p.topics.some((t) => t.includes(trimmed) || trimmed.includes(t))
    );
  }

  if (topics.length) {
    const topicSet = new Set(topics.map((t) => t.toLowerCase()));
    results = results.filter((p) => p.topics.some((t) => topicSet.has(t)));
  }

  return results
    .filter((p) => p.source === "scripture")
    .slice(0, limit)
    .map((p) => ({
      reference: p.reference,
      text: p.text,
      source: "curated_catalog" as const,
      catalog_passage_id: p.id,
    }));
}

export async function executeScriptureTool(
  name: string,
  args: Record<string, unknown>,
): Promise<{ passages: ScripturePassage[] }> {
  if (name === "lookup_passage") {
    const reference = String(args.reference ?? "");
    const passage = await lookupPassage(reference);
    return { passages: passage ? [passage] : [] };
  }

  if (name === "discover_passages") {
    const query = String(args.query ?? "");
    const topics = Array.isArray(args.topics)
      ? args.topics.map(String)
      : [];
    const limit = typeof args.limit === "number" ? args.limit : 3;
    return { passages: discoverPassages(query, topics, limit) };
  }

  return { passages: [] };
}

/** Skip tool lookup when client already prefetched the same reference. */
export function prefetchedCoversLookup(
  prefetched: ScripturePassage[] | undefined,
  reference: string,
): boolean {
  if (!prefetched?.length) return false;
  const parsed = parseBibleReference(reference);
  if (!parsed) return false;
  return prefetched.some(
    (p) => p.reference.toLowerCase() === parsed.displayReference.toLowerCase(),
  );
}
